src/data/sym/card.lean

```diff
@@ -36,9 +35,185 @@ Prove the general case of stars and bars.
 stars and bars
 -/
 
+def multichoose1 (n k : ℕ) := fintype.card (sym (fin n) k)
+
+def multichoose2 (n k : ℕ) := (n + k - 1).choose k
+
+def encode (n k : ℕ) (x : sym (fin n.succ) k.succ) : sym (fin n) k.succ ⊕ sym (fin n.succ) k :=
+if h : fin.last n ∈ x then
+  sum.inr ⟨x.val.erase (fin.last n), by { rw [multiset.card_erase_of_mem h, x.property], refl }⟩
+else begin
+  refine sum.inl ⟨x.val.map (λ a, ⟨if (a : ℕ) = n then 0 else a, _⟩), _⟩,
+  { split_ifs,
+    { rw pos_iff_ne_zero,
+      rintro rfl,
+      obtain ⟨w, r⟩ := @multiset.exists_mem_of_ne_zero _ x.val (λ h, by simpa [h] using x.property),
+      simpa [subsingleton.elim w 0] using r, },
+    { cases lt_or_eq_of_le (nat.le_of_lt_succ a.property); solve_by_elim } },
+  { rw [multiset.card_map, x.property] }
+end
+
+def decode (n k : ℕ) : sym (fin n) k.succ ⊕ sym (fin n.succ) k → sym (fin n.succ) k.succ
+| (sum.inl x) := ⟨x.val.map (λ a, ⟨a.val, a.property.step⟩),
+                  by simpa [multiset.card_map] using x.property⟩
```

Comment:
Can you extract everything to the right of the `:=` to a new definition called `sym.map`?